pkg/nsx/services/store.go

```diff
@@ -230,45 +216,93 @@ func transError(err error) error {
 	return err
 }
 
-func getAll(service *SecurityPolicyService, obj *v1alpha1.SecurityPolicy,
-	nsxSecurityPolicy *model.SecurityPolicy) ([]model.Group, *model.SecurityPolicy, []model.Rule, error) {
-	indexResults, err := service.GroupStore.ByIndex(util.TagScopeSecurityPolicyCRUID, string(obj.UID))
-	if err != nil {
-		return nil, nil, nil, err
-	}
-	var existingGroups []model.Group
-	for _, group := range indexResults {
-		existingGroups = append(existingGroups, group.(model.Group))
+func (service *SecurityPolicyService) OperateSecurityStore(sp *model.SecurityPolicy) error {
+	if sp == nil {
+		return nil
 	}
-	existingSecurityPolicy := model.SecurityPolicy{}
-	res, ok, err := service.SecurityPolicyStore.GetByKey(*nsxSecurityPolicy.Id)
-	if err != nil {
-		return nil, nil, nil, err
-	}
-	if ok {
-		existingSecurityPolicy = res.(model.SecurityPolicy)
-	}
-	indexResults, err = service.RuleStore.ByIndex(util.TagScopeSecurityPolicyCRUID, string(obj.UID))
-	if err != nil {
-		return nil, nil, nil, err
+	if sp.MarkedForDelete != nil && *sp.MarkedForDelete {
+		err := service.SecurityPolicyStore.Delete(*sp) // Pass in the object to be deleted, not the pointer
+		log.V(1).Info("delete security policy from store", "securitypolicy", sp)
+		if err != nil {
+			return err
+		}
+	} else {
+		err := service.SecurityPolicyStore.Add(*sp)
+		log.V(1).Info("add security policy to store", "securitypolicy", sp)
+		if err != nil {
+			return err
+		}
 	}
-	var existingRules []model.Rule
-	for _, rule := range indexResults {
-		existingRules = append(existingRules, rule.(model.Rule))
+	return nil
+}
+
+func (service *SecurityPolicyService) OperateRulesStore(sp *model.SecurityPolicy) error {
+	for _, rule := range sp.Rules {
+		if rule.MarkedForDelete != nil && *rule.MarkedForDelete {
+			err := service.RuleStore.Delete(rule)
+			log.V(1).Info("delete rule from store", "rule", rule)
+			if err != nil {
+				return err
+			}
+		} else {
+			err := service.RuleStore.Add(rule)
+			log.V(1).Info("add rule to store", "rule", rule)
+			if err != nil {
+				return err
+			}
+		}
 	}
-	return existingGroups, &existingSecurityPolicy, existingRules, nil
+	return nil
 }
 
-func (service *SecurityPolicyService) ListSecurityPolicyID() sets.String {
-	groups := service.GroupStore.ListIndexFuncValues(util.TagScopeSecurityPolicyCRUID)
-	groupSet := sets.NewString()
-	for _, group := range groups {
-		groupSet.Insert(group)
+func (service *SecurityPolicyService) OperateGroupsStore(gs *[]model.Group) error {
```

Comment:
ditto.
And why in `OperateSecurityStore` and `OperateRulesStore` the parameter is `sp *model.SecurityPolicy` but `OperateGroupsStore` didn't follow that?